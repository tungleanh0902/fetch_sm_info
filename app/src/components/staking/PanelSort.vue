<template>
  <tr class="panel-sort-container">
    <th>#</th>
    <th
      v-for="property in properties"
      :key="property.value"
      :class="{
        'sort-by': sort,
        'hide-xs': hideXs(property),
      }"
      class="panel-sort-table-header"
    >
      <a
        v-if="sort"
        v-tooltip.top="property.tooltip"
        class="sort-by-link"
        @click="orderBy(property.value)"
      >
        {{ property.title }}
        <i class="material-icons notranslate">arrow_drop_up</i>
      </a>
      <span v-else>{{ property.title }}</span>
    </th>
  </tr>
</template>

<script>
export default {
  name: `panel-sort`,
  props: {
    sort: {
      type: Object,
      default: null,
    },
    properties: {
      type: Array,
      required: true,
    },
    showOnMobile: {
      type: String,
      required: true,
    },
  },
  methods: {
    hideXs(property) {
      const primaryFields = ["smallName", "block_number"]
      const hideFieldMobile = property.value !== this.showOnMobile
      return hideFieldMobile && !primaryFields.includes(property.value)
    },
    orderBy(property) {
      const sortBys = this.$el.querySelectorAll(`.sort-by`)
      sortBys.forEach((el) => el.classList.remove(`active`, `desc`, `asc`))
      const index = this.properties.findIndex((p) => p.value === property)
      const el = sortBys[index]

      if (this.sort.property === property) {
        if (this.sort.order === `asc`) {
          this.sort.order = `desc`
        } else {
          this.sort.order = `asc`
        }
      } else {
        this.sort.property = property
      }
      if (this.sort.order === `asc`) {
        el.classList.add(`asc`)
      } else {
        el.classList.add(`desc`)
      }
      el.classList.add(`active`)
    },
  },
}
</script>

<style>
.panel-sort-container {
  padding: 1rem;
  border-bottom: 1px solid var(--bc-dim);
}

.panel-sort-table-header {
  font-size: var(--sm);
}

.sort-by i {
  font-size: var(--lg);
  position: relative;
  top: 6px;
  right: 4px;
}

.sort-by a {
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  color: var(--dim);
}

.sort-by a:hover {
  color: var(--link);
}

.sort-by.active a {
  color: var(--highlight);
}

.sort-by.asc i {
  color: var(--highlight);
}

.sort-by.desc i {
  transform: rotate(180deg);
  color: var(--highlight);
}
</style>
