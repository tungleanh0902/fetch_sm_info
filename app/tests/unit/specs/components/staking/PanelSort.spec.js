import { shallowMount, createLocalVue } from "@vue/test-utils"
import PanelSort from "src/components/staking/PanelSort"

describe(`PanelSort`, () => {
  let wrapper

  const localVue = createLocalVue()
  localVue.directive(`tooltip`, () => {})

  beforeEach(() => {
    wrapper = shallowMount(PanelSort, {
      localVue,
      propsData: {
        sort: {
          order: `asc`,
        },
        properties: [
          {
            value: `id`,
            title: `ID`,
          },
          {
            value: `amount`,
            title: `AMOUNT`,
          },
        ],
        showOnMobile: "amount",
      },
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show an arrow according to ordering`, () => {
    const firstCol = wrapper.element.querySelector(`.sort-by`)
    const link = wrapper.element.querySelector(`.sort-by-link`)
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
    link.click()
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
    expect(firstCol.className.split(` `)).toContain(`asc`)
    link.click()
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(firstCol.className.split(` `)).toContain(`desc`)
    link.click()
    expect(firstCol.className.split(` `)).toContain(`asc`)
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
  })

  it(`should only sort one col actively`, () => {
    const firstCol = wrapper.element.querySelectorAll(`.sort-by`)[0]
    const secondCol = wrapper.element.querySelectorAll(`.sort-by`)[1]
    const firstLink = wrapper.element.querySelectorAll(`.sort-by-link`)[0]
    const secondLink = wrapper.element.querySelectorAll(`.sort-by-link`)[1]
    firstLink.click()
    secondLink.click()
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(secondCol.className.split(` `)).toContain(`asc`)
  })

  describe("showOnMobile", () => {
    beforeEach(() => {
      wrapper = shallowMount(PanelSort, {
        localVue,
        propsData: {
          sort: {
            property: `block_number`,
            order: `asc`,
          },
          properties: [
            {
              value: `block_number`,
              title: `Block Number`,
            },
            {
              value: `amount`,
              title: `Amount`,
            },
            {
              value: `other`,
              title: `Other`,
            },
          ],
          showOnMobile: "amount",
        },
      })
    })

    it(`should show mobile field correctly`, () => {
      wrapper.setProps({ showOnMobile: "amount" })
      const numHidden = wrapper
        .findAll("th")
        .filter((w) => w.classes("hide-xs")).length
      expect(numHidden).toBe(1)
    })
  })
})
