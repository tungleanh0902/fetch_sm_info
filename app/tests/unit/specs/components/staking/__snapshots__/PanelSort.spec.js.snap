// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelSort has the expected html structure 1`] = `
<tr
  class="panel-sort-container"
>
  <th>
    #
  </th>
   
  <th
    class="panel-sort-table-header sort-by hide-xs"
  >
    <a
      class="sort-by-link"
    >
      
      ID
      
      <i
        class="material-icons notranslate"
      >
        arrow_drop_up
      </i>
    </a>
  </th>
  <th
    class="panel-sort-table-header sort-by"
  >
    <a
      class="sort-by-link"
    >
      
      AMOUNT
      
      <i
        class="material-icons notranslate"
      >
        arrow_drop_up
      </i>
    </a>
  </th>
</tr>
`;
